File: backend/services/gemini_service.py
from google import genai
from google.genai import types
import os
import logging
import json
from datetime import datetime

logger = logging.getLogger(__name__)

class GeminiService:
    def __init__(self):
        api_key = os.environ.get("GEMINI_API_KEY")
        if not api_key:
            logger.warning("GEMINI_API_KEY not found in environment variables")
            raise ValueError("GEMINI_API_KEY not set")
            
        self.client = genai.Client(api_key=api_key)
        self.model = os.environ.get("GEMINI_MODEL", "gemini-2.0-flash-exp") 
        logger.info(f"Initialized Gemini Service with model: {self.model}")

    def generate_chat_response(self, messages: list, context: str = None) -> str:
        try:
            gemini_messages = []
            
            for msg in messages:
                gemini_messages.append(types.Content(
                    role=msg.role if msg.role == 'user' else 'model',
                    parts=[types.Part(text=msg.content)]
                ))

            system_instruction = """
            You are Orbit AI, an advanced DeFi portfolio manager and assistant.
            Your goal is to help users manage their crypto investments, analyze risks, and find yield opportunities.
            Tone: Professional, futuristic, concise, and helpful. 
            Do not give financial advice, but provide data-driven insights.
            """
            
            if context:
                system_instruction += f"\n\nUser Context:\n{context}"

            response = self.client.models.generate_content(
                model=self.model,
                contents=gemini_messages,
                config=types.GenerateContentConfig(
                    system_instruction=system_instruction,
                    temperature=0.7
                )
            )
            
            return response.text
        except Exception as e:
            logger.error(f"Error generating chat response: {str(e)}")
            return "I'm having trouble connecting to the Orbit mainframe. Please try again later."

    def generate_weekly_analysis(self, portfolio_data: dict):
        try:
            prompt = f"""
            Analyze the following COMPLETE DeFi portfolio data (Loops, CEX Positions, Standard DeFi Positions) and generate a comprehensive weekly report.
            
            Portfolio Data:
            {json.dumps(portfolio_data, indent=2, default=str)}
            
            Your task is to populate the 'Risk Heatmap' and 'Orbit Intelligence Stream' widgets of the dashboard.
            
            OUTPUT REQUIREMENTS (Strict JSON):
            
            1. 'risk_metrics': An array of 5 objects for the Radar Chart.
               Each object must have: 
               - 'subject': string (Must be exactly these 5: 'Liquidation', 'Volatility', 'Protocol', 'Peg', 'Strategy')
               - 'A': integer (0-100, where 100 is perfectly safe/good, 0 is critical risk)
               - 'fullMark': integer (always 100)
               
            2. 'insights': An array of 3-5 insight objects.
               Each object must have:
               - 'type': string ('warning', 'opportunity', 'info')
               - 'title': string (Short, punchy header)
               - 'message': string (Actionable advice or observation)
               - 'impact': string (e.g. 'High Risk', '+$450/yr', 'Liquidity')
               
            3. 'risk_score': integer (0-100 overall safety score)
            
            4. 'summary': string (Brief executive summary)

            Make the analysis realistic based on the provided data. If data is sparse, make reasonable assumptions for a general crypto portfolio but be conservative.
            """
            
            response = self.client.models.generate_content(
                model=self.model,
                contents=prompt,
                config=types.GenerateContentConfig(
                    response_mime_type="application/json",
                    temperature=0.4
                )
            )
            
            return json.loads(response.text)
        except Exception as e:
            logger.error(f"Error generating analysis: {str(e)}")
            return {
                "risk_metrics": [],
                "insights": [],
                "risk_score": 0,
                "summary": "Analysis failed."
            }

# Singleton instance
try:
    gemini_service = GeminiService()
except Exception as e:
    logger.error(f"Failed to initialize GeminiService: {e}")
    gemini_service = None

File: frontend/server/src/services/geminiService.ts
import pino from "pino";
import { GoogleGenerativeAI } from "@google/generative-ai";
import dotenv from "dotenv";
import path from "path";

// Ensure environment variables are loaded even if server boot order changes
if (process.env.API_ENV_PATH) {
  dotenv.config({ path: process.env.API_ENV_PATH });
}
dotenv.config({ path: path.resolve(process.cwd(), "server/.env") });
dotenv.config();

const logger = pino({ level: process.env.LOG_LEVEL || "info" });

export class GeminiService {
  private client: GoogleGenerativeAI | null;
  private modelName: string;

  constructor() {
    const apiKey = process.env.GEMINI_API_KEY;
    if (!apiKey) {
      logger.error("GEMINI_API_KEY not set");
      this.client = null;
      this.modelName = "";
      return;
    }
    this.client = new GoogleGenerativeAI(apiKey);
    this.modelName = process.env.GEMINI_MODEL || "gemini-1.5-flash";
    logger.info({ model: this.modelName }, "Initialized Gemini Service");
  }

  async generateChatResponse(messages: { role: string; content: string }[], context?: string): Promise<string> {
    if (!this.client) return "Orbit AI is offline. Please try again later.";
    try {
      const model = this.client.getGenerativeModel({ model: this.modelName });
      const systemInstruction = [
        "You are Orbit AI, an advanced DeFi portfolio manager and assistant.",
        "Your goal is to help users manage their crypto investments, analyze risks, and find yield opportunities.",
        "Tone: Professional, futuristic, concise, and helpful.",
        "Do not give financial advice, but provide data-driven insights."
      ].join("\n");

      const history = messages.map(m => ({ role: m.role === "user" ? "user" : "model", parts: [{ text: m.content }] }));
      const finalParts = context ? [{ text: `${systemInstruction}\n\nUser Context:\n${context}` }] : [{ text: systemInstruction }];

      const res = await model.generateContent({ contents: [...history, { role: "user", parts: finalParts }] });
      const text = res.response.text();
      return text || "No response generated.";
    } catch (e) {
      logger.error({ err: e }, "Error generating chat response");
      return "I'm having trouble connecting to the Orbit mainframe. Please try again later.";
    }
  }

  async generateWeeklyAnalysis(portfolioData: Record<string, unknown>): Promise<Record<string, any>> {
    if (!this.client) return { risk_metrics: [], insights: [], risk_score: 0, summary: "Analysis failed." };
    try {
      const start = Date.now();
      const model = this.client.getGenerativeModel({ model: this.modelName });
      const prompt = `Analyze the following COMPLETE DeFi portfolio data (Loops, CEX Positions, Standard DeFi Positions) and generate a comprehensive weekly report.\n\nPopulate: Risk Heatmap + Orbit Intelligence Stream. Stick to the required fields only.\n\nPortfolio Data:\n${JSON.stringify(portfolioData, null, 2)}`;

      const generationConfig = {
        responseMimeType: "application/json",
        responseSchema: {
          type: "object",
          properties: {
            risk_metrics: {
              type: "array",
              items: {
                type: "object",
                properties: {
                  subject: { type: "string", enum: ["Liquidation", "Volatility", "Protocol", "Peg", "Strategy"] },
                  A: { type: "integer", minimum: 0, maximum: 100 },
                  fullMark: { type: "integer", minimum: 100, maximum: 100 }
                },
                required: ["subject", "A", "fullMark"]
              }
            },
            insights: {
              type: "array",
              items: {
                type: "object",
                properties: {
                  type: { type: "string", enum: ["warning", "opportunity", "info"] },
                  title: { type: "string" },
                  message: { type: "string" },
                  impact: { type: "string" }
                },
                required: ["type", "title", "message", "impact"]
              }
            },
            risk_score: { type: "integer", minimum: 0, maximum: 100 },
            summary: { type: "string" }
          },
          required: ["risk_metrics", "insights", "risk_score", "summary"]
        }
      } as any;

      const res = await model.generateContent({ contents: [{ role: "user", parts: [{ text: prompt }] }], generationConfig });
      let jsonText = res.response.text();
      const durationMs = Date.now() - start;
      logger.info({ durationMs, textLen: jsonText?.length || 0, sample: jsonText ? jsonText.slice(0, 180) : "" }, "Gemini weekly analysis raw response");
      try {
        // Some models still wrap JSON in markdown fences; strip them if present
        if (jsonText && jsonText.trim().startsWith("```")) {
          jsonText = jsonText.replace(/^```json\s*/i, "").replace(/^```\s*/i, "").replace(/```\s*$/i, "");
        }
        return JSON.parse(jsonText);
      } catch (parseErr) {
        logger.error({ err: parseErr, sample: jsonText ? jsonText.slice(0, 180) : "" }, "Failed to parse Gemini analysis JSON");
        return { risk_metrics: [], insights: [], risk_score: 0, summary: "Analysis failed." };
      }
    } catch (e) {
      logger.error({ err: e }, "Error generating analysis");
      return { risk_metrics: [], insights: [], risk_score: 0, summary: "Analysis failed." };
    }
  }
}

export const geminiService = new GeminiService();